File: dotnet/Pyrolite/Pickle/Objects/ByteArrayConstructor.cs
/* part of Pyrolite, by Irmen de Jong (irmen@razorvine.net) */

using System;
using System.Collections;
using System.Text;

namespace Razorvine.Pickle.Objects
{

/// <summary>
/// Creates byte arrays (byte[]). 
/// </summary>
class ByteArrayConstructor : IObjectConstructor {

	public object construct(object[] args) {
		// args for bytearray constructor: [ String string, String encoding ]
		// args for bytearray constructor (from python3 bytes): [ ArrayList ]
		if (args.Length != 1 && args.Length != 2)
			throw new PickleException("invalid pickle data for bytearray; expected 1 or 2 args, got "+args.Length);

		if(args.Length==1) {
			ArrayList values=(ArrayList) args[0];
			byte[] data=new byte[values.Count];
			for(int i=0; i<data.Length; ++i) {
				data[i] = Convert.ToByte(values[i]);
			}
			return data;
		} else {
			// This thing is fooling around with byte<>string mappings using an encoding.
			// I think that is fishy... but for now it seems what Python itself is also doing...
			string data = (string) args[0];
			string encoding = (string) args[1];
			if (encoding.StartsWith("latin-"))
				encoding = "ISO-8859-" + encoding.Substring(6);
			return Encoding.GetEncoding(encoding).GetBytes(data);
		}
	}
}

}

File: java/src/main/java/net/razorvine/pickle/objects/ByteArrayConstructor.java
package net.razorvine.pickle.objects;

import java.io.UnsupportedEncodingException;
import java.util.ArrayList;

import net.razorvine.pickle.IObjectConstructor;
import net.razorvine.pickle.PickleException;

/**
 * Creates byte arrays (byte[]).
 * 
 * @author Irmen de Jong (irmen@razorvine.net)
 */
public class ByteArrayConstructor implements IObjectConstructor {

	public Object construct(Object[] args) throws PickleException {
		// args for bytearray constructor: [ String string, String encoding ]
		// args for bytearray constructor (from python3 bytes): [ ArrayList<Number> ] or just [byte[]] (when it uses BINBYTES opcode)
		if (args.length != 1 && args.length != 2)
			throw new PickleException("invalid pickle data for bytearray; expected 1 or 2 args, got "+args.length);

		if(args.length==1) {
			if(args[0] instanceof byte[]) {
				return args[0];
			}
			@SuppressWarnings("unchecked")
			ArrayList<Number>values=(ArrayList<Number>)args[0];
			byte[] data=new byte[values.size()];
			for(int i=0; i<data.length; ++i) {
				data[i] = values.get(i).byteValue();
			}
			return data;
		} else {
			String data = (String) args[0];
			String encoding = (String) args[1];
			if (encoding.startsWith("latin-"))
				encoding = "ISO-8859-" + encoding.substring(6);
			try {
				return data.getBytes(encoding);
			} catch (UnsupportedEncodingException e) {
				throw new PickleException("error creating bytearray: " + e);
			}
		}
	}
}
